In [11]:
import os
import glob
import pandas as pd
from collections import OrderedDict, namedtuple
from plotly import tools

Load features computed for the predetermined sound samples.

In [2]:
path = '/mnt/data/Birdman/samples/features/features_petrels_bp1-8_wl256_th4.csv'
features = pd.read_csv(path, index_col=None)

In [3]:
selected_features = (
    ('petrel', 'presence of storm petrel'),
    ('sound.files', 'name of the file'),
    ('meanfreq', 'mean frequency (in kHz)'),
    ('sd', 'standard deviation of frequency'),
    ('freq.Q75', 'third quantile (in kHz)'),
    ('freq.IQR', 'interquantile range (in kHz)'),
    ('skew', 'skewness - asymmetry of the spectrum'),
    ('kurt', 'kurtosis - peakedness of the spectrum'),
    ('sp.ent', 'spectral entropy'),
    ('sfm', 'spectral flatness'),
    ('meanfun', 'average of fundamental frequency'),
    ('maxfun', 'maximum fundamental frequency'),
    ('meandom', 'average of dominant frequency'),
    ('dfrange', 'range of dominant frequency'),
    ('modindx', 'modulation index'),
    ('meanpeakf', 'mean peak frequency'))

feature_legend = OrderedDict(selected_features)
selected_features_names = [name for name, desc in selected_features]
features = features[selected_features_names]
features = features.fillna(0)

In [4]:
petrel_count = features['petrel'].value_counts()
print(petrel_count)

0    2482
1    1281
Name: petrel, dtype: int64


`1` denotes petrel, `0` lack of thereof (we'll call it *noise* for brevity). There is imbalance in favour of non-petrel features, primarily introduced with three files that in the selected regions represent only noise.

### Subsampling
STHELENA-02_20140605_* contain relatively few samples compared to the top 3. For training, we're going to take all sampels from STHELENA-02_20140605_* and subsample the top 3. It should be noted that top 3 has no petrels.

In [5]:
count_per_file = features['sound.files'].value_counts()
top3_names = list(count_per_file[:3].index)
rest_names = list(count_per_file[3:].index)
print(f'There are {len(features)} samples in total\n')
print(count_per_file)

There are 3763 samples in total

STHELENA-01_20140106_210000_0-15min.wav       840
STHELENA-02_20140108_210100_110-120min.wav    600
STHELENA-01_20140101_210000_55-105min.wav     600
STHELENA-02_20140605_200000_1.wav             209
STHELENA-02_20140605_200000_10.wav            185
STHELENA-02_20140605_200000_7.wav             178
STHELENA-02_20140605_200000_11.wav            164
STHELENA-02_20140605_200000_2.wav             156
STHELENA-02_20140605_200000_8.wav             138
STHELENA-02_20140605_200000_4.wav             135
STHELENA-02_20140605_200000_3.wav             134
STHELENA-02_20140605_200000_5.wav             132
STHELENA-02_20140605_200000_12.wav            104
STHELENA-02_20140605_200000_9.wav              89
STHELENA-02_20140605_200000_6.wav              51
STHELENA-02_20140605_200000_13.wav             48
Name: sound.files, dtype: int64


In [6]:
features_top3 = features[features['sound.files'].isin(top3_names)]
features_rest = features[features['sound.files'].isin(rest_names)]
noise_samples_top3 = sum(features_top3['petrel'] == 0)
noise_samples_rest = sum(features_rest['petrel'] == 0)
petrels_samples_top3 = sum(features_top3['petrel'] == 1)
petrels_samples_rest = sum(features_rest['petrel'] == 1)
assert(petrels_samples_top3 == 0) # There should be no petrel in the top 3
extra_noise_to_sample = petrels_samples_rest - noise_samples_rest
print(f'We need {extra_noise_to_sample} extra samples to have equal population of noise and petrel features')

We need 839 extra samples to have equal population of noise and petrel features


In [7]:
extra_noise_top3_df = features_top3.sample(extra_noise_to_sample, random_state=42)
features_top3_extra_removed_df = features_top3.drop(extra_noise_top3_df.index)
print(f'The leftover {len(features_top3_extra_removed_df)} will be used for testing purposes')

The leftover 1201 will be used for testing purposes


In [8]:
df = pd.concat([extra_noise_top3_df, features_rest]).reset_index()
df.head(5).T

0  \
index                                              1756   
petrel                                                0   
sound.files  STHELENA-02_20140108_210100_110-120min.wav   
meanfreq                                        3.60811   
sd                                              1.82977   
freq.Q75                                          5.202   
freq.IQR                                          3.314   
skew                                            1.89412   
kurt                                            9.50197   
sp.ent                                         0.979052   
sfm                                            0.816575   
meanfun                                         5.33333   
maxfun                                          5.33333   
meandom                                         1.27823   
dfrange                                           5.125   
modindx                                         6.14634   
meanpeakf                                       0.97619   

                                                      1  \
index                                              1440   
petrel                                                0   
sound.files  STHELENA-02_20140108_210100_110-120min.wav   
meanfreq                                        3.58382   
sd                                              1.82665   
freq.Q75                                          5.164   
freq.IQR                                          3.267   
skew                                            1.83392   
kurt                                            9.44598   
sp.ent                                         0.978609   
sfm                                            0.825397   
meanfun                                         5.33333   
maxfun                                          5.33333   
meandom                                         1.42473   
dfrange                                          5.0625   
modindx                                         11.4074   
meanpeakf                                       0.97619   

                                                      2  \
index                                              1508   
petrel                                                0   
sound.files  STHELENA-02_20140108_210100_110-120min.wav   
meanfreq                                        3.59702   
sd                                               1.8226   
freq.Q75                                          5.158   
freq.IQR                                          3.237   
skew                                            1.78846   
kurt                                            8.89957   
sp.ent                                         0.978441   
sfm                                            0.838956   
meanfun                                         5.33333   
maxfun                                          5.33333   
meandom                                         1.27895   
dfrange                                          3.5625   
modindx                                         8.05263   
meanpeakf                                       0.97619   

                                                     3  \
index                                             2407   
petrel                                               0   
sound.files  STHELENA-01_20140101_210000_55-105min.wav   
meanfreq                                       3.45414   
sd                                              1.7458   
freq.Q75                                         4.883   
freq.IQR                                         2.974   
skew                                           1.48091   
kurt                                            6.3322   
sp.ent                                        0.977618   
sfm                                           0.810595   
meanfun                                        4.74363   
maxfun                                         5.33333   
meandom                                        1.4573

Now that we see that results are OK, we can preprocess all the data. **One change though**.  We will concatenate the noise together and for ML training simply cut off 1201 samples from the bottom.

In [12]:
features_path = glob.glob('/mnt/data/Birdman/samples/features/raw/features*.csv')

In [13]:
for path in features_path:
    print(path)
    filename = os.path.splitext(os.path.basename(path))[0]
    features = pd.read_csv(path, index_col=None)
    features = features.fillna(0)
    count_per_file = features['sound.files'].value_counts()
    top3_names = list(count_per_file[:3].index)
    rest_names = list(count_per_file[3:].index)
    features_top3 = features[features['sound.files'].isin(top3_names)]
    features_rest = features[features['sound.files'].isin(rest_names)]
    noise_samples_top3 = sum(features_top3['petrel'] == 0)
    noise_samples_rest = sum(features_rest['petrel'] == 0)
    petrels_samples_top3 = sum(features_top3['petrel'] == 1)
    petrels_samples_rest = sum(features_rest['petrel'] == 1)
    assert(petrels_samples_top3 == 0) # There should be no petrel in the top 3
    extra_noise_to_sample = petrels_samples_rest - noise_samples_rest
    extra_noise_top3_df = features_top3.sample(extra_noise_to_sample, random_state=42)
    features_top3_extra_removed_df = features_top3.drop(extra_noise_top3_df.index)
    df = pd.concat([features_top3_extra_removed_df, extra_noise_top3_df, features_rest]).reset_index()
    df = df.drop('index', axis=1)
    df.to_csv(filename + '.csv', index=False)

/mnt/data/Birdman/samples/features/raw/features_petrels_bp1-6_wl256_th2.csv
/mnt/data/Birdman/samples/features/raw/features_petrels_bp1-8_wl256_th10.csv
/mnt/data/Birdman/samples/features/raw/features_petrels_bp1-8_wl256_th2.csv
/mnt/data/Birdman/samples/features/raw/features_petrels_bp1-8_wl256_th4.csv
/mnt/data/Birdman/samples/features/raw/features_petrels_bp1-12_wl256_th4.csv
